components/asset-store-controller-manager/pkg/assethook/engine/metadata_engine.go

```diff
@@ -0,0 +1,120 @@
+package engine
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"github.com/kyma-project/kyma/components/asset-store-controller-manager/pkg/apis/assetstore/v1alpha2"
+	"github.com/kyma-project/kyma/components/asset-store-controller-manager/pkg/assethook"
+	"github.com/kyma-project/kyma/components/asset-store-controller-manager/pkg/assethook/api/v1alpha1"
+	pkgPath "github.com/kyma-project/kyma/components/asset-store-controller-manager/pkg/path"
+	"github.com/pkg/errors"
+	"io"
+	"io/ioutil"
+	"mime/multipart"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+//go:generate mockery -name=MetadataExtractor -output=automock -outpkg=automock -case=underscore
+type MetadataExtractor interface {
+	Extract(ctx context.Context, object Accessor, basePath string, files []string, services []v1alpha2.WebhookService) ([]File, error)
+}
+
+type File struct {
+	Name     string
+	Metadata *json.RawMessage
+}
+
+type metadataEngine struct {
+	webhook    assethook.Webhook
+	timeout    time.Duration
+	fileReader func(filename string) ([]byte, error)
+}
+
+func NewMetadataExtractor(webhook assethook.Webhook, timeout time.Duration) MetadataExtractor {
+	return &metadataEngine{
+		webhook:    webhook,
+		timeout:    timeout,
+		fileReader: ioutil.ReadFile,
+	}
+}
+
+func (e *metadataEngine) Extract(ctx context.Context, object Accessor, basePath string, files []string, services []v1alpha2.WebhookService) ([]File, error) {
+	results := make(map[string]*json.RawMessage)
+	for _, service := range services {
+		filtered, err := pkgPath.Filter(files, service.Filter)
+		if err != nil {
+			return nil, errors.Wrapf(err, "while filtering files with regex %s", service.Filter)
+		}
+
+		body, contentType, err := e.buildQuery(basePath, filtered)
+		if err != nil {
+			return nil, errors.Wrap(err, "while building multipart query")
+		}
+
+		response := &v1alpha1.MetadataResponse{}
+		err = e.webhook.Do(ctx, contentType, service, body, response)
+		if err != nil {
+			return nil, errors.Wrap(err, "while sending request to metadata webhook")
+		}
+
+		results = e.mergeMetadata(results, response.Data)
+	}
+
+	return e.toFiles(results), nil
+}
+
+func (*metadataEngine) mergeMetadata(current map[string]*json.RawMessage, results []v1alpha1.MetadataResultSuccess) map[string]*json.RawMessage {
```

Comment:
`mergeMetadata` is for my a bad name, better is `replaceMetadata`. By `merge` developer can thinking about concatenating old fields with new fields, not replacing.